packages/react/src/utils/mergeReactProps.ts

```diff
@@ -1,36 +1,74 @@
 import * as React from 'react';
 import type { BaseUIEvent, WithBaseUIEvent } from './types';
 
+type MergableProps<T extends React.ElementType> =
+  | WithBaseUIEvent<React.ComponentPropsWithRef<T>>
+  | ((
+      otherProps: WithBaseUIEvent<React.ComponentPropsWithRef<T>>,
+    ) => WithBaseUIEvent<React.ComponentPropsWithRef<T>>)
+  | undefined;
+
 /**
- * Merges multiple sets of React props such that their event handlers are called in sequence
- * (the leftmost one being called first), and allows the user to prevent the subsequent event handlers from being
+ * Merges multiple sets of React props. It follows the Object.assign pattern where the leftmost object's fields overwrite
+ * the conflicting ones from others. This doesn't apply to event handlers, `className` and `style` props.
+ * Event handlers are merged such that they are called in sequence (the leftmost one being called first),
+ * and allows the user to prevent the subsequent event handlers from being
  * executed by attaching a `preventBaseUIHandler` method.
  * It also merges the `className` and `style` props, whereby the classes are concatenated
  * and the leftmost styles overwrite the subsequent ones.
+ *
+ * Props can either be provided as objects or as functions that take the previous props as an argument.
+ * The function will receive the merged props up to that point (going from right to left):
+ * so it the case of `(obj1, fn, obj2, obj3)`, `fn` will receive the merged props of `obj2` and `obj3`.
+ * The function is responsible for chaining event handlers if needed (i.e. we don't run the merge logic).
+ *
+ * Event handlers returned by the functions are not automatically prevented when `preventBaseUIHandler` is called.
+ * They must check `event.baseUIHandlerPrevented` themselves and bail out if it's true.
```

Comment:
Can you elaborate more on this? Will all event handlers specified in our internal prop getters for components work by default without needing a check?

Reply:
Whenever a function is passed to `mergeReactProps`, it will be called with the merged props and the `preventBaseUIHandler` is not taken into consideration. Most of our prop getter functions actually call `mergeReactProps` internally, with one notable exception: `useButton`. In its case, the logic for prop merging is more complex as it depends on the button's disabled state.

In the case of functions returned from Floating UI's `useInteractions`, they don't know anything about `preventBaseUIHandler`, so I'm calling them without parameters and provide the result to `mergeReactProps` so our merging logic is run on them. This alone should be sufficient to fix 1., but I refactored `mergeReactProps` to avoid having to nest prop getters and help prevent bugs like 2.